artemis-commons/src/main/java/org/apache/activemq/artemis/utils/BoundedOrderedExecutorFactory.java

```diff
@@ -0,0 +1,269 @@
+
+      static {
+         STATE_UPDATER = AtomicIntegerFieldUpdater.newUpdater(OrderedExecutorState.class, "state");
+      }
+
+      private volatile int state = 0;
+
+   }
+
+   private abstract static class BoundedOrderedExecutorL1Pad extends OrderedExecutorState {
+
+      protected long p01, p02, p03, p04, p05, p06, p07;
+      protected long p10, p11, p12, p13, p14, p15, p16, p17;
+   }
+
+   private static final class BoundedOrderedExecutor extends BoundedOrderedExecutorL1Pad implements BoundedExecutor {
+
+      private final Queue<Runnable> commands;
+      private final Runnable executeCommandsTask;
+      private final Executor delegate;
+      private final int maxBurstSize;
+      private final int capacity;
+
+      BoundedOrderedExecutor(Queue<Runnable> commands, int capacity, int maxBurstSize, Executor delegate) {
+         this.commands = commands;
+         this.executeCommandsTask = this::executeCommands;
+         this.delegate = delegate;
+         this.capacity = capacity;
+         this.maxBurstSize = maxBurstSize;
+      }
+
+      private boolean tryAcquire() {
+         //much cheaper than CAS when highly contended
+         final long oldState = STATE_UPDATER.getAndIncrement(this);
+         final boolean isAcquired = oldState == RELEASED;
+         return isAcquired;
+      }
+
+      private boolean isReleased() {
+         return STATE_UPDATER.get(this) == RELEASED;
+      }
+
+      private void release() {
+         //StoreStore + LoadStore: much cheaper than a volatile store
+         STATE_UPDATER.lazySet(this, RELEASED);
+      }
+
+      private void executeCommands() {
+         final Queue<Runnable> commands = this.commands;
+         final int maxBurstSize = this.maxBurstSize;
+         //let others consumers to try to acquire the lock and drain the tasks
+         while (!commands.isEmpty() && tryAcquire()) {
+            try {
+               drainCommands(commands, maxBurstSize);
+            } finally {
+               release();
+            }
+         }
+      }
+
+      @Override
+      public int capacity() {
+         return this.capacity;
+      }
+
+      @Override
+      public int pendingTasks() {
+         return this.commands.size();
+      }
+
+      @Override
+      public boolean isEmpty() {
+         return this.commands.isEmpty();
+      }
+
+      @Override
+      public boolean tryExecute(Runnable command) {
+         //no optimisations on recursive offers
+         if (commands.offer(command)) {
+            if (isReleased() && !commands.isEmpty()) {
+               this.delegate.execute(executeCommandsTask);
+            }
+            return true;
+         } else {
+            return false;
+         }
+      }
+
+      @Override
+      public void execute(Runnable command) {
+         //no optimisations on recursive offers
+         if (!commands.offer(command)) {
+            throw new RejectedExecutionException("can't submit the task: max capacity reached");
```

Comment:
I see where the word bounded came from.. there is a limit?

What about the Growable Queue?